python/Ganga/GPIDev/Adapters/IGangaFile.py

```diff
@@ -36,8 +40,44 @@ def location(self):
     def get(self):
         """
         Retrieves locally all files that were uploaded before that 
-        """
-        raise NotImplementedError
+        Order of priority about where a file is going to be placed are:
+            1) The localDir as defined in the schema
+            2) The Job outpudir of the parent job if the localDir is not defined
+            3) raise an exception if neither are defined correctly
+        """
+        if self.localDir:
+            if not os.path.isdir(self.localDir):
+                msg = "Folder '%s' doesn't exist. Please construct this before 'get'-ing a file." % self.localDir
+                logger.error(msg)
+                #os.makedirs(self.localDir)
+                raise GangaException(msg)
+            to_location = self.localDir
+        else:
+            should_raise = True
+            if self._getParent() is not None:
+                try:
+                    to_location = self.getJobObject().outputdir
+                    should_raise = False
+                except AssertionError:
+                    should_raise = True
+
+            if should_raise:
+                msg = "%s: Failed to get file object. Please set the `localDir` parameter and try again. e.g. file.localDir=os.getcwd();file.get()" % getName(self)
+                logger.error(msg)
+                logger.debug("localDir value: %s" % self.localDir)
+                logger.debug("parent: %s" % self._getParent() )
+                raise GangaException(msg)
+
+        if not os.path.isfile(os.path.join(to_location, self.namePattern)):
+            try:
+                return self.copyTo(to_location)
+            finally:
```

Comment:
I would say that if the `copyTo` raises an exception, then the `localDir` shouldn't be changed, i.e. the `localDir` only updated on a successful copy to the Ganga repo.